src/System.Text.Utf8String/System/Text/Utf8Span.cs

```diff
@@ -0,0 +1,782 @@
+        }
+
+        public bool Equals(string other)
+        {
+            CodePointEnumerator thisEnumerator = GetCodePointEnumerator();
+            Utf16LittleEndianCodePointEnumerator otherEnumerator = new Utf16LittleEndianCodePointEnumerator(other);
+
+            while (true)
+            {
+                bool hasNext = thisEnumerator.MoveNext();
+                if (hasNext != otherEnumerator.MoveNext())
+                {
+                    return false;
+                }
+
+                if (!hasNext)
+                {
+                    return true;
+                }
+
+                if (thisEnumerator.Current != otherEnumerator.Current)
+                {
+                    return false;
+                }
+            }
+        }
+
+        public static bool operator ==(Utf8Span left, Utf8Span right)
+        {
+            return left.Equals(right);
+        }
+
+        public static bool operator !=(Utf8Span left, Utf8Span right)
+        {
+            return !left.Equals(right);
+        }
+
+        public static bool operator ==(Utf8Span left, string right)
+        {
+            return left.Equals(right);
+        }
+
+        public static bool operator !=(Utf8Span left, string right)
+        {
+            return !left.Equals(right);
+        }
+
+        public static bool operator ==(string left, Utf8Span right)
+        {
+            return right.Equals(left);
+        }
+
+        public static bool operator !=(string left, Utf8Span right)
+        {
+            return !right.Equals(left);
+        }
+
+        public int CompareTo(Utf8Span other)
+        {
+            throw new NotImplementedException();
+        }
+
+        public int CompareTo(string other)
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="index">Index in UTF-8 code units (bytes)</param>
+        /// <returns>Length in UTF-8 code units (bytes)</returns>
+        public Utf8Span Substring(int index)
+        {
+            return Substring(index, Length - index);
+        }
+
+        /// <summary>
+        ///
+        /// </summary>
+        /// <param name="index">Index in UTF-8 code units (bytes)</param>
+        /// <returns>Length in UTF-8 code units (bytes)</returns>
+        public Utf8Span Substring(int index, int length)
+        {
+            if (index < 0)
+            {
+                throw new ArgumentOutOfRangeException("index");
+            }
+
+            if (length < 0)
+            {
+                // TODO: Should we support that?
+                throw new ArgumentOutOfRangeException("length");
+            }
+
+            if (length == 0)
+            {
+                return Empty;
+            }
+
+            if (length == Length)
+            {
+                return this;
+            }
+
+            if (index + length > Length)
```

Comment:
This check should happen before
`if (length == Length) return this;` (which I don't think is necessary).

Reply:
Potential overflow in addition; should subtract 
```
index > Length - length 
```